src/proxy/pool.rs

```diff
@@ -12,80 +12,539 @@
+        let (request_sender, connection) = builder
+            .handshake(::hyper_util::rt::TokioIo::new(tls_stream))
+            .await
+            .map_err(Error::HttpHandshake)?;
+
+        // spawn a task to poll the connection and drive the HTTP state
+        // if we got a drain for that connection, respect it in a race
+        // it is important to have a drain here, or this connection will never terminate
+        let mut driver_drain = self.state.timeout_rx.clone();
+        tokio::spawn(async move {
+            debug!("starting a connection driver for {:?}", clone_key);
+            tokio::select! {
+                    _ = driver_drain.changed() => {
+                        debug!("draining outer HBONE connection {:?}", clone_key);
+                    }
+                    res = connection=> {
+                        match res {
+                            Err(e) => {
+                                error!("Error in HBONE connection handshake: {:?}", e);
+                            }
+                            Ok(_) => {
+                                debug!("done with HBONE connection handshake: {:?}", res);
+                            }
+                        }
+                    }
+            }
+        });
 
-impl<F> hyper::rt::Executor<F> for TokioExec
-where
-    F: std::future::Future + Send + 'static,
-    F::Output: Send + 'static,
-{
-    fn execute(&self, fut: F) {
-        tokio::spawn(fut);
+        Ok(request_sender)
     }
 }
 
 #[derive(Debug, Clone)]
-struct Client(http2::SendRequest<Empty<Bytes>>);
+// A sort of faux-client, that represents a single checked-out 'request sender' which might
+// send requests over some underlying stream using some underlying http/2 client
+pub struct ConnClient {
+    sender: http2::SendRequest<Empty<Bytes>>,
+    stream_count: Arc<AtomicU16>, // the current streamcount for this client conn.
+    stream_count_max: u16,        // the max streamcount associated with this client.
+    // A WL key may have many clients, but every client has no more than one WL key
+    wl_key: WorkloadKey, // the WL key associated with this client.
+}
 
-impl Poolable for Client {
-    fn is_open(&self) -> bool {
-        self.0.is_ready()
+impl ConnClient {
+    pub fn at_max_streamcount(&self) -> bool {
+        let curr_count = self.stream_count.load(Ordering::Relaxed);
+        debug!("checking streamcount: {curr_count}");
+        if curr_count >= self.stream_count_max {
+            return true;
+        }
+        false
     }
 
-    fn reserve(self) -> Reservation<Self> {
-        let b = self.clone();
-        let a = self;
-        Reservation::Shared(a, b)
+    pub fn send_request(
+        &mut self,
+        req: Request<Empty<Bytes>>,
+    ) -> impl Future<Output = hyper::Result<Response<Incoming>>> {
+        // TODO should we enforce streamcount per-sent-request? This would be slow.
+        self.stream_count.fetch_add(1, Ordering::Relaxed);
```

Comment:
am I missing something or do we never decrement this?

BTW -- we need to decrement not when the send_request is done, since the streaming body persists beyond the `Response`

Reply:
We never do - once we hit the max we pop the conn out of the pool, start a new one, and let this one expire naturally.